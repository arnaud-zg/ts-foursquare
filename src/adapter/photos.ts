import { NPayload, NPhotos } from '../../types'
import { ECommonError } from '../constants/error'
import { adaptPayload } from './payload'

export const adaptGetPhotosDetails = (
  payload: NPayload.IPayload<NPhotos.IResponse>
): NPhotos.IPhoto => {
  const response = adaptPayload<NPhotos.IResponse>(payload)

  if (!response || !response.photo) {
    throw new Error(ECommonError.RESPONSE_NOT_VALID)
  }

  return response.photo
}
