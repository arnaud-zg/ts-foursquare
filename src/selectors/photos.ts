import { NPhotos, NStore } from '../../types'

export const photoSelector = (
  state: NStore.IState,
  listId: string
): NPhotos.IPhoto => photosSelector(state).entities[listId]

export const photosSelector = (state: NStore.IState) => state.photos
