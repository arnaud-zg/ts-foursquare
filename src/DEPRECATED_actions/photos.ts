import { ActionType, createAsyncAction } from 'typesafe-actions'
import { NRequest } from '../../types'
import { adaptGetPhotosDetails } from '../adapter'

export const getPhotosDetailsAsync = createAsyncAction(
  'GET_PHOTOS_DETAILS_REQUEST',
  'GET_PHOTOS_DETAILS_SUCCESS',
  'GET_PHOTOS_DETAILS_FAILURE',
  'GET_PHOTOS_DETAILS_CANCEL'
)<
  NRequest.IPhotosDetailsPayload,
  ReturnType<typeof adaptGetPhotosDetails>,
  Error,
  string
>()

export const photosActions = {
  getPhotosDetailsAsync,
}

export type TPhotosAction = ActionType<typeof photosActions>
