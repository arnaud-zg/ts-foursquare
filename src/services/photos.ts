import { fromFetch } from 'rxjs/fetch'
import { catchError, map, switchMap } from 'rxjs/operators'
import { NRequest } from '../../types'
import { adaptGetPhotosDetails } from '../adapter'
import { EApiDefaultParameters, EApiPathnames } from '../constants/api'
import { IStandaloneConfig } from '../standalone'
import { generatePath } from '../utils/generatePath'
import { getLocationHref } from '../utils/url'
import { processFetchError, processFetchResponse } from './fetch'

export const getObservablePhotosDetails = ({
  payload,
}: {
  config?: IStandaloneConfig
  payload: NRequest.IPhotosDetailsPayload
}) => {
  const { photoId } = payload

  return fromFetch(
    getLocationHref({
      origin: EApiDefaultParameters.ORIGIN,
      pathname: generatePath(EApiPathnames.PHOTOS_DETAILS, { photoId }),
      param: {},
    })
  ).pipe(
    switchMap(processFetchResponse),
    map(adaptGetPhotosDetails),
    catchError(processFetchError)
  )
}
