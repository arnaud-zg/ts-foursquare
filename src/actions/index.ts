import { TLifeAction } from './life'
import { TListsAction } from './lists'
import { TPhotosAction } from './photos'
import { TVenuesAction } from './venues'

export type TRootAction =
  | TLifeAction
  | TListsAction
  | TPhotosAction
  | TVenuesAction

export * from './life'
export * from './lists'
export * from './photos'
export * from './venues'
