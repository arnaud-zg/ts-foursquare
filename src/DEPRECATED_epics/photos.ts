import { Epic } from 'redux-observable'
import { of } from 'rxjs/internal/observable/of'
import { catchError, filter, map, switchMap, takeUntil } from 'rxjs/operators'
import { isActionOf } from 'typesafe-actions'
import { NStore } from '../../types'
import { getPhotosDetailsAsync, TRootAction } from '../DEPRECATED_actions'
import { adaptGetPhotosDetails } from '../adapter'
import { getObservablePhotosDetails } from '../services'

export const getPhotosDetailsEpic: Epic<
  TRootAction,
  TRootAction,
  NStore.IState
> = action$ =>
  action$.pipe(
    filter(isActionOf(getPhotosDetailsAsync.request)),
    switchMap(action =>
      getObservablePhotosDetails({ action }).pipe(
        map(adaptGetPhotosDetails),
        map(getPhotosDetailsAsync.success),
        catchError(err => of(getPhotosDetailsAsync.failure(err))),
        takeUntil(
          action$.pipe(filter(isActionOf(getPhotosDetailsAsync.cancel)))
        )
      )
    )
  )
