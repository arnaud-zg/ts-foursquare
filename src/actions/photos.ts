import { NRequest } from '../../types'
import { getObservablePhotosDetails } from '../services/photos'
import { IStandaloneConfig } from '../standalone'

export interface GetPhotosDetailsProps {
  config?: IStandaloneConfig
  payload: NRequest.IPhotosDetailsPayload
}

export const getPhotosDetails = ({ config, payload }: GetPhotosDetailsProps) =>
  getObservablePhotosDetails({ config, payload }).toPromise()
