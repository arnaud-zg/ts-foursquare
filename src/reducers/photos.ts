import { createReducer, getType } from 'typesafe-actions'
import { NPhotosState } from '../../types'
import { getPhotosDetailsAsync, TPhotosAction } from '../actions'

export const initialState: NPhotosState.IState = {
  entities: {},
}

export const photosReducer = createReducer<NPhotosState.IState, TPhotosAction>(
  initialState
).handleAction(
  getType(getPhotosDetailsAsync.success),
  (state, action): NPhotosState.IState => {
    return {
      ...state,
      entities: {
        ...state.entities,
        [action.payload.id]: action.payload,
      },
    }
  }
)
