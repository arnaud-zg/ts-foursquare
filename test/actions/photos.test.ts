import { getPhotosDetailsAsync } from '../../src/actions/photos'

describe('actions/photos', () => {
  it(`should make a snapshot of getPhotosDetailsAsync`, () => {
    expect(getPhotosDetailsAsync).toMatchSnapshot()
  })
})
