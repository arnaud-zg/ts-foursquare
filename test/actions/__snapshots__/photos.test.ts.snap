// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions/photos should make a snapshot of getPhotosDetailsAsync 1`] = `
Object {
  "cancel": [Function],
  "failure": [Function],
  "request": [Function],
  "success": [Function],
}
`;
